routemaster/server/tests/test_endpoints.py

```diff
@@ -326,6 +326,39 @@ def test_create_label_409_for_deleted_label(client, create_label):
     assert response.status_code == 409
 
 
+def test_restore_label(app, client, create_deleted_label):
+    create_deleted_label('foo', 'test_machine', {'foo': 'bar'})
+
+    label_metadata = {'bar': 'baz'}
+    response = client.post(
+        '/state-machines/test_machine/labels/foo/restore_and_restart',
+        data=json.dumps({
+            'metadata': label_metadata,
+        }),
+        content_type='application/json',
+    )
+
+    assert response.status_code == 201
+    assert response.json['metadata'] == label_metadata
+
+    with app.new_session():
+        label = app.session.query(Label).one()
+        assert label.metadata == label_metadata
+        assert label.deleted is False
+
+
+def test_restore_unknown_label(client):
+    response = client.post(
+        '/state-machines/test_machine/labels/foo/restore_and_restart',
+        data=json.dumps({
+            'metadata': {},
+        }),
+        content_type='application/json',
+    )
+
+    assert response.status_code == 404
```

Comment:
Maybe also assert that a label of the given name was not created? (This would help clarify the difference between this endpoint and `create_label`)